File: app/src/main/java/com/stochanskyi/itunesloader/presentation/base/BaseFragment.java
package com.stochanskyi.itunesloader.presentation.base;

import android.content.Context;
import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.stochanskyi.itunesloader.presentation.activities.mainActivity.navigation.Screen;
import com.stochanskyi.itunesloader.presentation.base.MVP.BasePresenter;
import com.stochanskyi.itunesloader.presentation.base.MVP.IBaseView;

public abstract class BaseFragment<P extends BasePresenter<V>, V extends IBaseView> extends Fragment
        implements IBaseView {
    protected P presenter;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        createPresenterInstance();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        view.setOnTouchListener((View v, MotionEvent event) -> true);
        initViews(view);
        initPresenter();
        presenter.onStart();
    }
}

File: app/src/main/java/com/stochanskyi/itunescharts/presentation/base/mvp/BaseFragment.kt
package com.stochanskyi.itunescharts.presentation.base.mvp

import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment

abstract class BaseFragment<T : IBasePresenter<out IBaseFragment> > :
    Fragment(),
    IBaseFragment {

    abstract val presenter: T

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initPresenter()
        initViews()
        presenter.onStart()
    }

    abstract fun initPresenter()

    override fun close() {
        parentFragment?.let {
            if (it.childFragmentManager.backStackEntryCount > 0) {
                it.childFragmentManager.popBackStack()
            }
        } ?: activity?.let {
            if (it.supportFragmentManager.backStackEntryCount > 0) {
                it.supportFragmentManager.popBackStack()
            }
        }
    }
}